blockchain/v2/scheduler.go

```diff
@@ -0,0 +1,614 @@
+	}
+
+	if sc.allBlocksProcessed() {
+		return scFinishedEv{}, nil
+	}
+
+	return noOp, nil
+}
+
+func (sc *scheduler) handleBlockProcessError(event pcBlockVerificationFailure) (Event, error) {
+	// TODO
+	return noOp, nil
+}
+
+func (sc *scheduler) handleAddNewPeer(event addNewPeer) (Event, error) {
+	err := sc.addPeer(event.peerID)
+	if err != nil {
+		return scSchedulerFail{reason: err}, nil
+	}
+	return noOp, nil
+}
+
+// XXX: unify types peerError
+func (sc *scheduler) handlePeerError(event peerError) (Event, error) {
+	err := sc.removePeer(event.peerID)
+	if err != nil {
+		return scSchedulerFail{reason: err}, nil // xxx: bit extreme
+		// AZ: not clear where the peerError event is coming from, I presume is
+		// from the switch. It is possible that the removePeer fails here for legitimate reasons
+		// for example if a peer timeout or error was handled just before this.
+	}
+	return noOp, nil
+}
+
+func (sc *scheduler) handleTryPrunePeer(event tryPrunePeer) (Event, error) {
+	prunablePeers := sc.prunablePeers(sc.peerTimeout, sc.minRecvRate, event.time)
+
+	if len(prunablePeers) == 0 {
+		return noOp, nil
+	}
+
+	peerID := prunablePeers[0]
+	err := sc.removePeer(peerID)
+	if err != nil {
+		return scSchedulerFail{reason: err}, nil // xxx: should be fatal
+	}
+
+	return scPeerPruned{peerID: peerID}, nil
+}
+
+// AZ: This seems to schedule a single block request, is this the intent?
+func (sc *scheduler) handleTrySchedule(event trySchedule) (Event, error) {
+	pendingBlocks := sc.numBlockInState(blockStatePending)
+	receivedBlocks := sc.numBlockInState(blockStateReceived)
+	todo := math.Min(float64(sc.targetPending-pendingBlocks), float64(sc.targetReceived-receivedBlocks))
+	// AZ: I think it's better to have a single sc.maxBlockPoolSize to indicate the total number of blocks
+	// either pending or received and not processed and then have something like:
+	// todo := sc.maxBlockPoolSize-pendingBlocks-receivedBlocks
+	if todo > 0 {
+		height := sc.minHeight()
+		// AZ: this may return math.MaxInt64 if all blocks are processed and then getStateAtHeight() will panic
+		if sc.getStateAtHeight(height) == blockStateNew {
+			// Get all peers with height >= height that are active
+			allPeers := sc.getPeersAtHeight(height)
+			bestPeer := sc.selectPeer(allPeers) // XXX: maybe this should return a p2p.ID
+			err := sc.markPending(bestPeer.peerID, height, event.time)
+			if err != nil {
+				return scSchedulerFail{reason: err}, nil // XXX: peerError might be more appropriate
+			}
+			return scBlockRequest{peerID: bestPeer.peerID, height: height}, nil
+		}
+	}
+
+	return noOp, nil
+}
+
+func (sc *scheduler) handleStatusResponse(event bcStatusResponse) (Event, error) {
+	err := sc.setPeerHeight(event.peerID, event.height)
+	if err != nil {
+		return scPeerError{peerID: event.peerID, reason: err}, nil
+	}
+	return noOp, nil
+}
+
+func (sc *scheduler) handle(event Event) (Event, error) {
```

Comment:
Where are the incoming `blockRequest` and `statusRequest` handled? If by the scheduler, it must live after fastsync is finished.

Reply:
They would come directly from the reactors `Receive` method.